### PR TITLE
Reduce ram usage by 25%, improve startup performance, make Greeter prettier.

### PR DESCRIPTION
**Pull request by UniversalSuperBox**
This PR resolves the single greatest user of RAM by percentage, the wallpaper. It also resolves the single heaviest cause of startup lag by percentage, the thing that checks all of the user's desired wallpapers to make sure they actually load. And it resolves the position of the Greeter wallpaper.

That's right, three types of resolution.

Resolving the Greeter wallpaper means positioning it correctly. Have you ever noticed that the wallpaper jiggles a bit when you unlock your device? That's because the Greeter's wallpaper image is actually shorter than the one that serves as your background image. The first commit here makes sure the Greeter's image and the one that serves as the Stage's background match. I took extra care to make sure that the clock and InfoGraphic didn't move on a narrow (phone) greeter. The InfoGraphic and login box moved a bit on the wide greeter... they're actually in the vertical center of the screen now.

Next we optimize the biggest RAM hog in Lomiri by setting sourceSize on its wallpapers to an appropriate value. You can get a whole blog post about why that's needed and why it's done the way it is in the second commit's message. But for the stats: This saves 60MB of RAM when Lomiri is run at 540x960 (the resolution of the BQ E4.5) with software rendering and with the default wallpaper. That was ~25% of its memory usage at that resolution. The savings go up if your wallpaper is larger than the default wallpaper's beefy 4096x2304. They go down if your device has a higher resolution than 540x960 or if your wallpaper is smaller than the default. But in all cases, *it's free RAM*. This commit required special care to make sure that all images were loaded at the same resolution. Not doing so would incur multiple image loads from disk and squander our saved RAM.

The final commit starts by reducing Shell creation time by about 10%. How that translates into real-world startup performance I'm not sure, but it saves about a minute in the tests on CI. We do that by first optimizing WallpaperResolver. It loaded the final image in its candidates to make sure it could, but it returned the final image whether or not the load was successful. So there was no reason to load it in the first place. Then, we make all of the Images between the Wallpaper and the WallpaperResolver asynchronous. This means that the shell no longer freezes when it's changing wallpapers, it starts up just a little bit faster, and the Greeter appears sooner when requested. It had an unfortunate side effect, though: the wallpaper would be transparent for a moment and then pop in. I fixed that by adding a small fade-in animation that plays after the image loads. When I showed it to people, they asked me "what changed?" so I think the UX is good enough.

The final commit could probably be split into 3, but I don't think there's any reason to. They would conflict with each other on revert anyway.

### CODE REVIEW

File: qml/Components/Wallpaper.qml
Comment (mardy) [inline qml/Components/Wallpaper.qml]: I'm not sure whether it's needed, but just to be on the safe side I'd initialize this to `0.0`

File: qml/Components/WallpaperResolver.qml
Comment (mardy) [inline qml/Components/WallpaperResolver.qml]: does this actually happen? It seems an impossible case.

File: qml/Components/Wallpaper.qml
Comment (mardy) [inline qml/Components/Wallpaper.qml:86]: If you used a delayed binding for the source property, you could get rid of this.

**Review by mardy — APPROVED** [review]
Thanks for the tests!!

I added a few comments inline, but nothing that should prevent this from being merged.

File: qml/Components/Wallpaper.qml
Comment (UniversalSuperBox) [reply qml/Components/Wallpaper.qml:86]: I'm not sure I understand. What am I replacing with a delayed binding?

If you mean the timer, it's there because the Wallpaper will just *not load* if its sourceSize is unset. I know that it'll bite me or someone else working on Lomiri in the future (because it's such a dumb non-obvious effect) and so I added the warning.

File: qml/Components/WallpaperResolver.qml
Comment (UniversalSuperBox) [reply qml/Components/WallpaperResolver.qml]: This may have been a relic from when I was trying to use `root.candidates.slice(0, -1)` instead of `repeater.count` for this for loop. That actually caused some unexpected behavior in QML, the binding for `background` stopped firing when objects inside the repeater changed. I went back to `repeater.count` and it fixed the issue.

I'll fix it up.

File: qml/Components/Wallpaper.qml
Comment (UniversalSuperBox) [inline qml/Components/Wallpaper.qml:70]: I've been thinking about this more. Since the wallpaper loading is now asynchronous, we could save more memory by passing sourceSize as a `size` based on the current orientation of the shell. The image would have to reload every time the user rotated their phone, but it wouldn't cause a UI thread pause now that the images are async.

I haven't benchmarked it to see how much memory we'd save, but in theory it could be worth it. Maybe I should wait until after we merge this though? It'd be pretty easy to do, just replacing `real sourceSize` / `real backgroundSourceSize` with `size` values instead.

Comment (UniversalSuperBox) [reply qml/Components/Wallpaper.qml:70]: I'll try it out later. This is a large enough RAM savings to warrant doing first without waiting for more development.

Comment (mardy) [reply qml/Components/Wallpaper.qml:70]: I wouldn't even try it. I'm not convinced that you'll get any savings there, and you risk making the rotation animation slower. I think this is very good the way it is :-)

File: qml/Components/Wallpaper.qml
Comment (mardy) [reply qml/Components/Wallpaper.qml:86]: I meant that instead of using a Timer here, you could have the `source` changed signal connected with a delayed binding, and at that point in the handler of this binding you could be sure that the sourceSize must have been set. If it isn't set, then you'd print the error message. But it's a minor thing, so do it only if you like: I'm approving this anyways.

**Review by mardy — APPROVED** [review]

**Comment by lduboeuf** [comment]
humm, when having the default wallpaper, i see the screen turn grey-ish just before PIN code screen

**Comment by UniversalSuperBox** [comment]
When? Could you provide a screenshot and steps to reproduce?

**Comment by lduboeuf** [comment]
hard to take a screen shot, but after boot and after the "enter pin code" notification and just before the PIN code screen , the screen goes grey for 1 second.
I have that kind of background which is the default one if i'm not wrong:
screenshot20210609_150054732

**Comment by UniversalSuperBox** [comment]
Yes, this is the same as all Wallpapers. When they need to reload, they start as gray then fade in once the image is ready. That means that the first Wallpaper you see in Lomiri (be it the PIN unlock or the cover page) will be gray for a moment then fade in.

**Pull request merged**
